File: back-end/src/main/java/com/example/backend/service/AttendanceService.java
package com.example.backend.service;

import com.example.backend.entities.Attendance;
import com.example.backend.service.base.BaseService;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

public interface AttendanceService extends BaseService<Attendance> {
    List<Attendance> findByStudentId(Long studentId);
    List<Attendance> findByTripId(Long tripId);
    Optional<Attendance> findByIdWithDetails(Long attendanceId);
    List<Attendance> findStudentAttendanceInPeriod(Long studentId, LocalDateTime start, LocalDateTime end);
    List<Attendance> findByTripAndStatus(Long tripId, String status);
    List<Attendance> findUnnotifiedAttendance(LocalDateTime cutoffTime);
    long countTodaysPresentAttendance(Long schoolId);
    List<Attendance> findSchoolAttendanceInPeriod(Long schoolId, LocalDateTime start, LocalDateTime end);
    List<Attendance> findByParentId(Long parentId);
    void recordAttendance(Long studentId, Long tripId, String status, String notes);
    void updateAttendanceStatus(Long attendanceId, String status, String notes);
    void markAsNotified(Long attendanceId);
    List<Attendance> findAttendanceWithStats(Long schoolId, LocalDateTime start, LocalDateTime end);
    /**
     * Find all attendance records for a school
     *
     * @param schoolId the ID of the school
     * @return list of attendance records
     */
    List<Attendance> findAllSchoolAttendance(Long schoolId);
} 

File: front-end/src/services/AttendanceService.ts
import { Attendance } from '../core/entities/attendance.entity';

export interface AttendanceService {
  getAllAttendance(): Promise<Attendance[]>;
  getAttendanceById(id: number): Promise<Attendance>;
  getAttendanceWithDetails(id: number): Promise<Attendance>;
  findByStudentId(studentId: number): Promise<Attendance[]>;
  findByTripId(tripId: number): Promise<Attendance[]>;
  findStudentAttendanceInPeriod(studentId: number, start: string, end: string): Promise<Attendance[]>;
  findByTripAndStatus(tripId: number, status: string): Promise<Attendance[]>;
  findUnnotifiedAttendance(cutoffTime: string): Promise<Attendance[]>;
  countTodaysPresentAttendance(schoolId: number): Promise<number>;
  findSchoolAttendanceInPeriod(schoolId: number, start: string, end: string): Promise<Attendance[]>;
  findByParentId(parentId: number): Promise<Attendance[]>;
  recordAttendance(attendanceData: any): Promise<void>;
  updateAttendanceStatus(attendanceId: number, status: string, notes?: string): Promise<void>;
  markAsNotified(attendanceId: number): Promise<void>;
  findAttendanceWithStats(schoolId: number, start: string, end: string): Promise<Attendance[]>;
} 